### PR TITLE
LightningDataModule to load GeoTIFF files

### PR DESCRIPTION
**Pull request by weiji14**
## What I am changing

- A [LightningDataModule](https://lightning.ai/docs/pytorch/2.1.0/data/datamodule.html) to load GeoTIFF data

## How I did it

- Using `torchdata` to construct the DataPipe
- GeoTIFF files are read using `rasterio`
- Train/validation split is 80%/20%

TODO:
- [x] Install `torchdata` dependency
- [x] Initial implementation of GeoTIFFDataPipeModule
- [x] Add extra parameters to control DataLoader (e.g. num_workers)
- [x] Add unit tests
- [ ] Refactor to load GeoTIFF data from s3 bucket instead of local drive
- [ ] etc

Notes:
- Have tried using `rioxarray` to read the GeoTIFFs, but seems a little slower than `rasterio`
- Also experimented with loading from NetCDF files using xarray's h5netcdf engine (about same speed as rioxarray loading GeoTIFF)
- Fastest seems to be loading from Zarr, but would require re-formatting of data, so leaving that to a future PR.

## How you can test it

- Download the GeoTIFF files from the s3 bucket (TODO add instructions)
- Run `python trainer.py fit --trainer.max_epochs=20 --trainer.precision=16-mixed --data.data_path=data --data.batch_size=32 --data.num_workers=8` locally

## Related Issues

References:
- https://zen3geo.readthedocs.io/en/v0.6.2/walkthrough.html

### CODE REVIEW

File: src/datamodule.py [hunk 1]
```diff
@@ -1,33 +1,33 @@
...
     """
-    Torch Dataset that returns tensors of size (13, 256, 256) with random
-    values.
+    Read a GeoTIFF file using rasterio into a numpy.ndarray, and convert it
+    to a torch.Tensor (float16 dtype).
     """
+    # GeoTIFF - Rasterio
+    with rasterio.open(fp=filepath) as dataset:
+        array: np.ndarray = dataset.read()
+        tensor: torch.Tensor = torch.as_tensor(data=array.astype(dtype="float16"))
```
Comment (yellowcap) [inline src/datamodule.py:21]: Is float32 to float16 a save tansformation?

File: src/datamodule.py [hunk 2]
```diff
@@ -1,33 +1,33 @@
...
-class RandomDataset(torch.utils.data.Dataset):
+def _array_to_torch(filepath: str) -> torch.Tensor:
     """
-    Torch Dataset that returns tensors of size (13, 256, 256) with random
-    values.
+    Read a GeoTIFF file using rasterio into a numpy.ndarray, and convert it
+    to a torch.Tensor (float16 dtype).
     """
+    # GeoTIFF - Rasterio
+    with rasterio.open(fp=filepath) as dataset:
```
Comment (yellowcap) [inline src/datamodule.py:19]: Did you experiment with other file openers? Maybe the loader gets more stable if we use 
```
from skimage import io
im = io.imread(filepath)
```

or other tif specific loaders like

https://pypi.org/project/tifffile/

Could be worth a shot to see if that helps stabilizing the loader when compared to zarr.

File: src/datamodule.py (hunk 1 above)
Comment (weiji14) [reply src/datamodule.py:21]: There will be some loss of floating point precision, but we'll likely be using 16-bit precision training (see https://lightning.ai/docs/pytorch/2.1.0/common/precision_intermediate.html) to speed up the model training, so best to pre-emptively convert the data to float16 dtype here.

File: src/datamodule.py (hunk 2 above)
Comment (weiji14) [reply src/datamodule.py:19]: So `skimage`'s imread actually uses `tifffile` behind the scenes for reading TIFF files, see https://github.com/scikit-image/scikit-image/blob/441fe68b95a86d4ae2a351311a0c39a4232b6521/skimage/io/_io.py#L16-L68. I'll need to see if skimage/tifffile supports reading from s3 buckets directly like rasterio/GDAL. Found this thread https://github.com/cgohlke/tifffile/issues/125 which looks interesting.

**Comment by weiji14** [comment]
> * [ ]  Refactor to load GeoTIFF data from s3 bucket instead of local drive

Decided to handle reading from s3 in a separate PR, because it was about 10x slower than reading from a local disk, even from the same us-east-1 region. Specifically, a mini-batch took about 0.2it/s when reading from s3, compared to ~2it/s from a local data folder. Might need to play with some I/O or networking related settings.

**Comment by weiji14** [comment]
Again, merging directly in the interest of speed. Will refactor to try other drivers (following discussion at https://github.com/Clay-foundation/model/pull/52#discussion_r1404277273) later.
